### PR TITLE
Advanced Building Blocks - Enumerable

### PR DESCRIPTION
**Pull request by MkrtichSargsyan**
#  Advanced Building Blocks - Enumerable

In this milestone, we:

 - we created enumerable methods within the enumerable module
 - we test our methods with multiply_els method

### CODE REVIEW

File: .github/workflows/linters.yml [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+    runs-on: ubuntu-18.04
+    steps:
+      - uses: actions/checkout@v2
+      - uses: actions/setup-ruby@v1
+        with:
+          ruby-version: 2.6.x
+      - name: Setup Rubocop
+        run: |
+          gem install --no-document rubocop:'~>0.81.0' # https://docs.rubocop.org/en/stable/installation/
+          [ -f .rubocop.yml ] || wget https://raw.githubusercontent.com/microverseinc/linters-config/master/ruby/.rubocop.yml
```
Comment (majovanilla) [inline .github/workflows/linters.yml:18]: You are not using the same `.rubocop.yml` file that Microverse requires. Please add the correct one to the root directory.

File: main.rb [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+  def my_each
+    ind = 0
+
+    while ind != length
+      yield self[ind]
+      ind += 1
+    end
+
+    self
+  end
```
Comment (majovanilla) [inline main.rb]: - [x] should have working `my_each` identical and returns the same thing as ruby's [`each`](https://ruby-doc.org/core-2.6.4/Array.html#method-i-each). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_each.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
- Your method should return an enumerator when no block is given.
- It should work with ranges.
- It should work with hashes.

File: main.rb [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+
+  def my_each_with_index
+    index = 0
+    my_each do |el|
+      yield(el, index)
+      index += 1
+    end
+
+    self
+  end
```
Comment (majovanilla) [inline main.rb]: - [x] should have working `my_each_with_index` identical and returns the same thing as ruby's [`each_with_index`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-each_with_index). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_each_with_index.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
- Your method should return an enumerator when no block is given.
- It should work with ranges.
- It should work with hashes.

File: main.rb [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+  def my_select
+    filtered_arr = []
+
+    while length != 0
+      yield(first) && filtered_arr.push(first)
+      shift
+    end
+
+    filtered_arr
+  end
```
Comment (majovanilla) [inline main.rb]: - [x] should have working `my_select` identical and returns the same thing as ruby's [`select`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-select). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_select.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
- Your method is not returning an enumerator when no block is given.
- It should return an array with all the elements that meet the given condition (it's returning more).
- It should work with ranges.
- It shouldn't mutate the original array.

File: main.rb [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+    while length != 0
+      if !yield(first)
+        output = false
+        break
+      else
+        shift
+      end
+    end
+    output
+  end
```
Comment (majovanilla) [inline main.rb]: - [x] should have working `my_all?` *identical and returns the same thing as ruby's [`all?`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-all-3F). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_all.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))

- It shouldn't modify the original array.
- Your method doesn't consider the possibility when no block or argument is given. It should return true if the enumerable contains truthy values and false if it contains at least one falsy value.
- Your method doesn't consider the possibility when a class (String, Numeric, etc) is given as an argument `(some_enumerable.my_all?(String)).`
- Your method doesn't consider the possibility when a Regex is given as an argument `(some_enumerable.my_all?(/w/)).`
- Your method doesn't consider the possibility when a number is given as an argument `(some_enumerable.my_all?(5)).`

File: main.rb [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+  def my_any?
+    output = false
+    my_each do |el|
+      if yield el
+        output = true
+        break
+      end
+    end
+    output
+  end
```
Comment (majovanilla) [inline main.rb]: - [ ] should have working `my_any?` identical and returns the same thing as ruby's [`any?`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-any-3F). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_any.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))

- The method is not allowing you to pass any arguments (a word, for example). Please fix that.

File: main.rb [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+    while length != 0
+      if yield first
+        output = false
+        break
+      else
+        shift
+      end
+    end
+    output
+  end
```
Comment (majovanilla) [inline main.rb]: - [x] should have working `my_none?` identical and returns the same thing as ruby's [`none?`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-none-3F). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_none.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))

- Your method doesn't consider the possibility when no block or argument is given. It should return true if the enumerable contains no truthy value.
- It should work with ranges.
- Your method doesn't consider the possibility when a class is given as an argument `(some_enumerable.my_none?(String)).`
- Your method doesn't consider the possibility when a Regex is given as an argument `(some_enumerable.my_none?(/w/)).`
- Your method doesn't consider the possibility when a number is given as an argument `(some_enumerable.my_none?(8)).`

File: main.rb [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+  # my_count
+
+  def my_count
+    count = 0
+
+    my_each do |el|
+      yield(el) && count += 1
+    end
+    count
+  end
```
Comment (majovanilla) [inline main.rb]: - [x] should have working `my_count` identical and returns the same thing as ruby's [`count`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-count). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_count.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))

- Your method doesn't consider the possibility when no argument or block is given. It should return the count of all elements inside the enumerable.
- Your method doesn't consider the possibility when a value is given as an argument `(some_enumerable.my_count(5)).` It should return the count of all elements that match the argument.
- It should work with ranges.

File: main.rb [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+
+  # my_map
+
+  def my_map(&proc_block)
+    new_arr = []
+
+    my_each { |el| new_arr.push(proc_block.call(el)) }
+
+    new_arr
+  end
```
Comment (majovanilla) [inline main.rb]: - [x] should have working `my_map` **(covered by automated tests)** identical and returns the same thing as ruby's [`map`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-map). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_map.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))

- Your method should return an enumerator if no block or argument is given.
- It should work with ranges.
- It should execute only the proc when both a block and a proc are given (for each element).

File: main.rb [hunk 10]
```diff
@@ -0,0 +1,122 @@
...
+
+  def my_inject(num)
+    output = num
+
+    my_each do |sum|
+      output = yield(output, sum)
+    end
+    output
+  end
+end
```
Comment (majovanilla) [inline main.rb]: - [x] should have working `my_inject` identical and returns the same thing as ruby's [`inject`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-inject). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_inject.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))

- It should raise an LocalJumpError when no block is given. 
- It should take a block as an argument and apply it to each element of the object. Your's it's not recognizing it. 
- Your method should consider the possibility when an initial value is given.
- Your method should work with symbols as arguments (:+) and perform the right operation.
- Your method should work with ranges.

**Review by majovanilla — CHANGES_REQUESTED** [review]
# Changes required :mag: 


Hey team, 👋

## Good job on this project! 👍

- Good Readme file
- Good Pull Request Message

### There are still some changes 👇 you need to implement to approve this request 🔍

### Required Changes: ✍️

- Please take a look at the comments I left in your code.
- You are using `shift` in some methods, that modifies the original array and you shouldn't.


## Please, make all REQUIRED changes and submit your request again! 😀

If you have any questions, feel free to send me a message through Slack (@majo)
I would be happy to help! 👩‍💻 


Happy coding 🤓

--Maria Reyes

**Review by andresporras3423 — CHANGES_REQUESTED** [review]
### Status: Changes Required :recycle:
**Hello Mike Sargsyan & Chad Van Wyk** 
Nice job with the improvements to your project :muscle:  :muscle: 
We still need to fix some issues. Make sure your enumerable methods can pass the next scenarios:
- [x] enumerables #my_all? when no block or argument is given returns true when none of the collection members are false or nil
- [x] enumerables #my_all? when no block or argument is given returns false when one of the collection members are false or nil
- [x] enumerables #my_all? when a Regex is passed as an argument returns true if all of the collection matches the Regex
- [x] enumerables #my_all? when a pattern other than Regex or a Class is given returns true if all of the collection matches the pattern
- [x] enumerables #my_any? when no block or argument is given returns true if at least one of the collection is not false or nil
- [x] enumerables #my_any? when a Regex is passed as an argument returns true if any of the collection matches the Regex
- [x] enumerables #my_none? when no block or argument is given returns true only if none of the collection members is true
- [x] enumerables #my_none? when a Regex is passed as an argument returns false if any of the collection matches the Regex
- [x] enumerables #my_inject when a block is given without an initail value combines all elements of enum by applying a binary operation, specified by a block::range
- [x] enumerables #my_inject when a block is given with an initail value combines all elements of enum by applying a binary operation, specified by a block
- [x] enumerables #my_inject when a symbol is specified without an initail value combines each element of the collection by applying the symbol as a named method
- [x] enumerables #my_inject when a symbol is specified without an initail value combines each element of the collection by applying the symbol as a named method:range
- [x] enumerables #my_inject when a symbol is specified with an initail value combines each element of the collection by applying the symbol as a named method
- [x] enumerables #my_inject when a symbol is specified with an initail value combines each element of the collection by applying the symbol as a named method::range

Please request another review once you are ready! :+1: 

Happy coding :relaxed:

Oscar Andrés Russi Porras

**Review by andresporras3423 — CHANGES_REQUESTED** [review]
Project review by Oscar Andrés Russi Porras (TSE)

**Review by andresporras3423 — CHANGES_REQUESTED** [review]
### Status: Changes Required :recycle:
**Hello Mike Sargsyan & Chad Van Wyk** 
Nice job with the improvements since the last review muscle:  :muscle: 
We still need to fix some issues with the inject method:
- [x] enumerables #my_inject when a block is given without an initail value combines all elements of enum by applying a binary operation, specified by a block::range
`Range.new(1, 5).my_inject { |prod, n| prod * n } == Range.new(1, 5).inject { |prod, n| prod * n }`
- [x] enumerables #my_inject when a block is given with an initial value combines all elements of enum by applying a binary operation, specified by a block
`Range.new(1, 4).my_inject(5) { |prod, n| prod * n } == Range.new(1, 4).inject(5) { |prod, n| prod * n }`
- [x] enumerables #my_inject when a symbol is specified without an initial value combines each element of the collection by applying the symbol as a named method
`[1,2,3,4,5].my_inject(:+) == [1,2,3,4,5].inject(:+)`
- [x] enumerables #my_inject when a symbol is specified without an initial value combines each element of the collection by applying the symbol as a named method:range
`Range.new(1, 5).my_inject(:*) == Range.new(1, 5).inject(:*)`
- [x] enumerables #my_inject when a symbol is specified with an initial value combines each element of the collection by applying the symbol as a named method
`[1,2,3,4].my_inject(5, :*) == [1,2,3,4].inject(5, :*)`
- [x] enumerables #my_inject when a symbol is specified with an initial value combines each element of the collection by applying the symbol as a named method::range
`Range.new(1, 4).my_inject(5, :*) == Range.new(1, 4).inject(5, :*)`

Please request another review once you are ready! :+1: 

Happy coding :relaxed:

Oscar Andrés Russi Porras

**Review by sumancrest0001 — APPROVED** [review]
**STATUS: APPROVED :+1: :100: 
Congratulation TEAM :clap:** 
You have passed all the tests and all the methods are working as expected. Now, you can merge the PR and submit the completion form.
**Happy coding !!!**

**Pull request merged**